src/librustdoc/clean/mod.rs

```diff
@@ -1298,31 +1292,32 @@ fn clean_qpath(hir_ty: &hir::Ty<'_>, cx: &mut DocContext<'_>) -> Type {
             let segments = if p.is_global() { &p.segments[1..] } else { &p.segments };
             let trait_segments = &segments[..segments.len() - 1];
             let trait_def = cx.tcx.associated_item(p.res.def_id()).container.id();
-            let trait_path = self::Path {
-                global: p.is_global(),
+            let trait_ = self::Path {
                 res: Res::Def(DefKind::Trait, trait_def),
                 segments: trait_segments.clean(cx),
             };
+            register_res(cx, trait_.res);
             Type::QPath {
                 name: p.segments.last().expect("segments were empty").ident.name,
                 self_def_id: Some(DefId::local(qself.hir_id.owner.local_def_index)),
                 self_type: Box::new(qself.clean(cx)),
-                trait_: Box::new(resolve_type(cx, trait_path, hir_id)),
+                trait_,
             }
         }
         hir::QPath::TypeRelative(ref qself, ref segment) => {
             let ty = hir_ty_to_ty(cx.tcx, hir_ty);
             let res = if let ty::Projection(proj) = ty.kind() {
                 Res::Def(DefKind::Trait, proj.trait_ref(cx.tcx).def_id)
             } else {
-                Res::Err
+                bug!("clean: expected associated item")
```

Comment:
I'll note that this panics on the following code:

```rust
#![feature(inherent_associated_types)]

pub struct Struct;

impl Struct {
    pub type AssocTy = usize;
    pub const AssocConst: Self::AssocTy = 42;
}
```

```
warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> inherent-associated-types.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information

error[E0223]: ambiguous associated type
 --> inherent-associated-types.rs:7:27
  |
7 |     pub const AssocConst: Self::AssocTy = 42;
  |                           ^^^^^^^^^^^^^ help: use fully-qualified syntax: `<Struct as Trait>::AssocTy`

error: internal compiler error: src/librustdoc/clean/mod.rs:1312:17: clean: expected associated item

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1147:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (517c28e42 2021-08-26) running on x86_64-apple-darwin

query stack during panic:
end of query stack
error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0223`.
```

However:

* A rustc error is emitted for this code, before rustdoc ICEs (I'm not very familiar with `inherent_associated_types`, so I'm not sure how the error would be fixed)
* This is an unstable and incomplete feature
* rustdoc panics on nightly too, though at a different spot (because this change caused rustdoc to fail early):

```
warning: the feature `inherent_associated_types` is incomplete and may not be safe to use and/or cause compiler crashes
 --> inherent-associated-types.rs:1:12
  |
1 | #![feature(inherent_associated_types)]
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(incomplete_features)]` on by default
  = note: see issue #8995 <https://github.com/rust-lang/rust/issues/8995> for more information

error[E0223]: ambiguous associated type
 --> inherent-associated-types.rs:7:27
  |
7 |     pub const AssocConst: Self::AssocTy = 42;
  |                           ^^^^^^^^^^^^^ help: use fully-qualified syntax: `<Struct as Trait>::AssocTy`

thread 'rustc' panicked at 'attempted .def_id() on invalid res: Err', /rustc/5eacec9ec7e112a0de1011519a57c45586d58414/compiler/rustc_hir/src/def.rs:538:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.56.0-nightly (5eacec9ec 2021-08-28) running on x86_64-apple-darwin

query stack during panic:
end of query stack
error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0223`.
```

Reply:
Ok, I opened a new PR for that: #88573. I'll drop the commit in this PR that changes the `Res::Err` to a `bug!` since I incorporated it into the other PR.